### ISSUE TITLE
grammar parser has problems

### ISSUE THREAD

**Issue by aeschli** [Microsoft/vscode-textmate#50]
_From @jens1o on June 22, 2017 15:24_

- VSCode Version: Code - Insiders 1.14.0-insider (a49a42680a4269b79c4e80f6d754bbded03a0593, 2017-06-21T05:15:42.843Z)
- OS Version: Windows_NT ia32 10.0.15063
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
path-intellisense|chr|1.4.2
gitlens|eam|4.1.3
tslint|eg2|0.15.0
LogFileHighlighter|emi|1.1.1
vscode-reveal|evi|0.0.9
php-debug|fel|1.10.0
php-intellisense|fel|1.4.1
auto-close-tag|for|0.4.2
gc-excelviewer|Gra|1.1.15
GitHubIssues|Hoo|0.1.2
composer|ika|0.5.0
smarty|imp|0.2.0
json-to-ts|Mar|1.4.3
vscode-apache|mrm|1.1.1
cpptools|ms-|0.11.4
php-docblocker|nei|1.2.0
vscode-versionlens|pfl|0.19.1
vscode-icons|rob|7.9.0
sharecode|Rol|0.4.1
code-spell-checker|str|1.2.0
vscode-todo-highlight|way|0.5.5
highlight-trailing-white-spaces|yba|0.0.2

(1 theme extensions excluded)

---

see https://github.com/atom/language-php/issues/221

It works in Atom, but not in vscode. :/

_Copied from original issue: Microsoft/vscode#29271_

**Comment by aeschli** [comment]
_From @jens1o on June 22, 2017 15:26_

I'm curious. This bot can say there is a bug?

**Comment by aeschli** [comment]
_From @jens1o on July 8, 2017 21:0_

A little ping here. So far no reaction, and it's really annoying, to say that in a nice way. :(

**Comment by aeschli** [comment]
_From @roblourens on July 9, 2017 3:45_

If I understand correctly, the same syntax file works in Atom, but not vscode - we are not adding `support.class` and `keyword.operator.class` to the class names on static member accesses. Does that imply there's a bug in vscode? Thoughts @aeschli? How do I debug this?

**Comment by aeschli** [comment]
_From @jens1o on July 9, 2017 6:17_

Yes, you're right. And from my point of view, this is an annoying bug where the stable users have to live with for one month.... (Since there was a huge grammar cleanup coming with 1.14, but it is not ready for vscode yet...)

So they might have think, wow, grammar update, I get better syntax highlighting... But vscode does not do that what atom does(the grammar is now awesome) and so we can't use it...

**Comment by aeschli** [comment]
Hmm, I also don't have an easy way to debug this. I try to understand the grammar, but didn't figure out how the "support.class.php" is getting pushed to the stack.

**Comment by aeschli** [comment]
_From @jens1o on July 11, 2017 14:45_

Maybe you can ask @50Wliu kindly and he will elaborate? :)

**Comment by aeschli** [comment]
_From @50Wliu on July 11, 2017 14:52_

Scope resolution happens [here](https://github.com/atom/language-php/blob/5c0ef02e2615c9ca2ecd33708356aa43f2409f5d/grammars/php.cson#L2296).  The first rule deals with the part before the `::`, all the other rules deal with the part after.  [`class-name`](https://github.com/atom/language-php/blob/5c0ef02e2615c9ca2ecd33708356aa43f2409f5d/grammars/php.cson#L318) (called [here](https://github.com/atom/language-php/blob/5c0ef02e2615c9ca2ecd33708356aa43f2409f5d/grammars/php.cson#L2308)) adds the `support.class.php` scope.

Also, @jens1o you seem to be excellent at timing your comments to be right before I open up my notifications :P.  This is like the fourth time I've opened up my notifications and the first one is a comment from you < 2 minutes ago.

**Comment by aeschli** [comment]
The problem is with the following rule:
```
				{
					"begin": "(?=[\\\\a-zA-Z_])",
					"end": "(?i)\\b([a-z_][a-z_0-9]*)?(?=[^a-z0-9_])",
					"endCaptures": {
						"1": {
							"name": "support.class.php"
						}
					},
					"patterns": [
						{
							"include": "#namespace"
						}
					]
				}
```
https://github.com/Microsoft/vscode/blob/master/extensions/php/syntaxes/php.tmLanguage.json#L309

Removing the positive lookahead (?=[^a-z0-9_])) lets the endCapture match.

**Comment by roblourens** [comment]
Maybe https://github.com/Microsoft/vscode/issues/30448 is also related?

**Comment by roblourens** [comment]
@aeschli are you saying that the regex is wrong, or that there's a bug in vscode-textmate? There are a lot of new issues about PHP syntax highlighting, some of which might just be due to grammar changes, but I wonder whether the grammar is being interpreted incorrectly in some cases.

**Comment by aeschli** [comment]
I'd say it is a bug in our engine. No idea why the positive lookahead prevents the match here.
But it's also possible that I didn't realize which other rule prevented the match. IMO it would be good if we have a debug mode where we can trace how our engine tries to match.

**Comment by roblourens** [comment]
Filed https://github.com/atom/language-php/issues/235, this is actually a grammar bug. It seems that Atom is not using the latest version of the grammar?

**Comment by 50Wliu** [comment]
@roblourens Microsoft/vscode#30448 is a grammar bug, but we have [a substantial amount of specs](https://github.com/atom/language-php/blob/e1290265f3d68316347e0ab2665686016b4b24b7/spec/php-spec.coffee#L616) covering the original issue reported by @jens1o, so it would appear that this one is a vscode-textmate bug.

**Comment by roblourens** [comment]
You're right, I confused it with https://github.com/Microsoft/vscode/issues/30448, too many PHP highlighting bugs around right now.

**Comment by jens1o** [comment]
Any progress right now, I'm seeing this error everyday? It's a shame to see this every day, as I code everyday and everywhere...

**Comment by roblourens** [comment]
I need to spend time debugging this, but I have a lot going on this week and am out next week, so I don't know when I'll have time.

**Comment by jens1o** [comment]
Okay, at least I know this issue is not that unimportant to you :)

Have fun next week!

**Comment by alexandrudima** [comment]
I can also try looking into this, but @aeschli @roblourens can you please distil a self-contained bug report that links to grammars and code snippets, but using immutable grammar links.

i.e. https://github.com/Microsoft/vscode/blob/master/extensions/php/syntaxes/php.tmLanguage.json#L309 is not an immutable grammar link because line 309 now contains a `}`.

> Tip:  (press <kbd>y</kbd> when viewing the file on github to generate an immutable link and never paste links in with branch names in them)

**Comment by mirao** [comment]
It seems to be fixed in 1.18. At least the issue #30531 has gone.

**Comment by jens1o** [comment]
finally :)
